Userland/Libraries/LibWeb/FileAPI/FileReader.cpp

```diff
@@ -0,0 +1,359 @@
+/*
+ * Copyright (c) 2023, Shannon Booth <shannon@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Assertions.h>
+#include <AK/Base64.h>
+#include <AK/ByteBuffer.h>
+#include <LibJS/Heap/Heap.h>
+#include <LibJS/Runtime/Promise.h>
+#include <LibJS/Runtime/Realm.h>
+#include <LibJS/Runtime/TypedArray.h>
+#include <LibTextCodec/Decoder.h>
+#include <LibWeb/Bindings/Intrinsics.h>
+#include <LibWeb/DOM/Event.h>
+#include <LibWeb/DOM/EventTarget.h>
+#include <LibWeb/FileAPI/Blob.h>
+#include <LibWeb/FileAPI/FileReader.h>
+#include <LibWeb/HTML/EventLoop/EventLoop.h>
+#include <LibWeb/HTML/EventNames.h>
+#include <LibWeb/HTML/Scripting/TemporaryExecutionContext.h>
+#include <LibWeb/MimeSniff/MimeType.h>
+#include <LibWeb/Platform/EventLoopPlugin.h>
+#include <LibWeb/Streams/AbstractOperations.h>
+#include <LibWeb/Streams/ReadableStream.h>
+#include <LibWeb/Streams/ReadableStreamDefaultReader.h>
+#include <LibWeb/WebIDL/DOMException.h>
+#include <LibWeb/WebIDL/ExceptionOr.h>
+
+namespace Web::FileAPI {
+
+FileReader::~FileReader() = default;
+
+FileReader::FileReader(JS::Realm& realm)
+    : DOM::EventTarget(realm)
+{
+}
+
+void FileReader::initialize(JS::Realm& realm)
+{
+    DOM::EventTarget::initialize(realm);
+    set_prototype(&Bindings::ensure_web_prototype<Bindings::FileReaderPrototype>(realm, "FileReader"));
+}
+
+void FileReader::visit_edges(JS::Cell::Visitor& visitor)
+{
+    visitor.visit(m_error);
+}
+
+JS::NonnullGCPtr<FileReader> FileReader::create(JS::Realm& realm)
+{
+    return realm.heap().allocate<FileReader>(realm, realm);
+}
+
+JS::NonnullGCPtr<FileReader> FileReader::construct_impl(JS::Realm& realm)
+{
+    return FileReader::create(realm);
+}
+
+// https://w3c.github.io/FileAPI/#blob-package-data
+ErrorOr<FileReader::Result> FileReader::blob_package_data(JS::Realm& realm, ByteBuffer bytes, Type type, Optional<String> const& mime_type, Optional<String> const& encoding_name)
+{
+    // A Blob has an associated package data algorithm, given bytes, a type, a optional mimeType, and a optional encodingName, which switches on type and runs the associated steps:
+    switch (type) {
+    case Type::DataURL:
+        // Return bytes as a DataURL [RFC2397] subject to the considerations below:
+        // Use mimeType as part of the Data URL if it is available in keeping with the Data URL specification [RFC2397].
+        // If mimeType is not available return a Data URL without a media-type. [RFC2397].
+        return MUST(AK::URL::create_with_data(mime_type.value_or(String {}), MUST(encode_base64(bytes)), true).to_string());
+    case Type::BinaryString:
+        // FIXME: Return bytes as a binary string, in which every byte is represented by a code unit of equal value [0..255].
```

Comment:
Return an `Error` here? Otherwise this will fall through into the `Type::Text` code below.

Reply:
The fallthrough was on purpose since from my testing it seemed to be similar-ish, but yeah, I think just erroring out would be more appropriate here.